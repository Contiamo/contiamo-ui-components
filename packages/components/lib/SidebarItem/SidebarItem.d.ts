/// <reference types="react" />
import * as React from "react";
import { Css } from "../types";
export interface Props {
    id?: string;
    css?: Css;
    className?: string;
    children?: React.ReactNode;
    onClick?: () => void;
    /** Disabled */
    disabled?: boolean;
    /** Active */
    active?: boolean;
}
declare const SidebarItem: (props: Props) => JSX.Element;
export default SidebarItem;
