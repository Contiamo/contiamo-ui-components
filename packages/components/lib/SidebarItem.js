"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var glamorous_1 = require("glamorous");
var utils_1 = require("@operational/utils");
var Container = glamorous_1.default.div(function (_a) {
    var theme = _a.theme, isDisabled = _a.isDisabled, isActive = _a.isActive;
    return (__assign({ label: "sidebarlink", backgroundColor: theme.colors.white }, theme.typography.body, { fontWeight: isActive ? 600 : 400, position: "relative", display: "flex", padding: theme.spacing / 3 + "px " + theme.spacing + "px", cursor: "pointer", textDecoration: "none", color: isActive ? theme.colors.linkText : theme.colors.gray80, opacity: isDisabled ? 0.25 : 1, "&:not(:first-child)": {
            borderTop: "1px solid",
            borderColor: theme.colors.secondarySeparator
        }, ":hover": {
            backgroundColor: utils_1.darken(theme.colors.white)(2)
        }, ":focus": {
            outline: 0
        } }));
});
exports.default = function (props) { return (React.createElement(Container, { key: props.id, css: props.css, onClick: props.onClick, className: props.className, isActive: !!props.active, isDisabled: !!props.disabled }, props.children)); };
//# sourceMappingURL=SidebarItem.js.map