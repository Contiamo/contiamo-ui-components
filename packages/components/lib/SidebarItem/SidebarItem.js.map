{"version":3,"file":"SidebarItem.js","sourceRoot":"","sources":["../../src/SidebarItem/SidebarItem.tsx"],"names":[],"mappings":";;;;;;;;;;AAAA,6BAA8B;AAC9B,uCAA8D;AAkB9D,IAAM,SAAS,GAAG,mBAAS,CAAC,GAAG,CAC7B,UAAC,EAAyF;QAAvF,gBAAK,EAAE,0BAAU,EAAE,sBAAQ;IAA4E,OAAA,YACxG,KAAK,EAAE,aAAa,EACpB,eAAe,EAAE,KAAK,CAAC,MAAM,CAAC,KAAK,EACnC,MAAM,EAAE,EAAE,IACP,KAAK,CAAC,UAAU,CAAC,IAAI,IACxB,UAAU,EAAE,QAAQ,CAAC,CAAC,CAAC,GAAG,CAAC,CAAC,CAAC,GAAG,EAChC,QAAQ,EAAE,UAAU,EACpB,YAAY,EAAE,WAAW,EACzB,WAAW,EAAE,KAAK,CAAC,MAAM,CAAC,SAAS,EACnC,OAAO,EAAE,MAAM,EACf,UAAU,EAAE,QAAQ,EACpB,cAAc,EAAE,eAAe,EAC/B,OAAO,EAAE,SAAO,KAAK,CAAC,OAAO,OAAI,EACjC,MAAM,EAAE,SAAS,EACjB,cAAc,EAAE,MAAM,EACtB,KAAK,EAAE,QAAQ,CAAC,CAAC,CAAC,KAAK,CAAC,MAAM,CAAC,QAAQ,CAAC,CAAC,CAAC,KAAK,CAAC,MAAM,CAAC,IAAI,IACxD,UAAU,CAAC,CAAC,CAAC,EAAE,OAAO,EAAE,IAAI,EAAE,aAAa,EAAE,MAAM,EAAE,CAAC,CAAC,CAAC,EAAE,IAC7D,QAAQ,EAAE;YACR,eAAe,EAAE,KAAK,CAAC,MAAM,CAAC,iBAAiB;SAChD,EACD,QAAQ,EAAE;YACR,OAAO,EAAE,CAAC;SACX,IACD;AAvBwG,CAuBxG,CACH,CAAA;AAED,IAAM,WAAW,GAAG,UAAC,KAAY,IAAK,OAAA,CACpC,oBAAC,SAAS,IACR,EAAE,EAAE,KAAK,CAAC,EAAE,EACZ,GAAG,EAAE,KAAK,CAAC,GAAG,EACd,OAAO,EAAE,KAAK,CAAC,OAAO,EACtB,SAAS,EAAE,KAAK,CAAC,SAAS,EAC1B,QAAQ,EAAE,CAAC,CAAC,KAAK,CAAC,MAAM,EACxB,UAAU,EAAE,CAAC,CAAC,KAAK,CAAC,QAAQ,IAE3B,KAAK,CAAC,QAAQ,CACL,CACb,EAXqC,CAWrC,CAAA;AAED,kBAAe,WAAW,CAAA"}