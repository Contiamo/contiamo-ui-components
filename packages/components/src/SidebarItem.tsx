import * as React from "react"
import glamorous, { GlamorousComponent } from "glamorous"
import { Theme } from "@operational/theme"
import { spin } from "@operational/utils"

export interface IProps {
  id?: string | number
  style?: {}
  className?: string
  label?: string
  children?: any
  open?: boolean
  onClick?: () => Promise<any>
  tooltip?: string
}

export interface IState {
  open: boolean
  updating: boolean // async, woo!
}

const style = ({ theme, children }: { theme: Theme; children?: Node }): {} => {
  // If we have children, style a caret.
  const caret: {} = children
    ? {
        content: '""',
        display: "block",
        width: 0,
        height: 0,
        marginLeft: "auto",
        border: "4px solid transparent",
        borderLeftColor: theme.colors.gray20,
        transition: ".15s transform ease"
      }
    : {}

  return {
    label: "sidebaritem",
    position: "relative",
    color: theme.colors.emphasizedText,

    "& .header": {
      position: "relative",
      display: "flex",
      alignItems: "center",
      padding: `${theme.spacing * 2 / 3}px ${theme.spacing}px`,
      borderTop: "1px solid",
      borderTopColor: theme.colors.gray10,
      cursor: "pointer",
      outline: "none",
      backgroundColor: theme.colors.white
    },

    "& .header:hover": {
      backgroundColor: theme.colors.gray10
    },

    "&.open .header": {
      borderBottom: "1px solid",
      borderBottomColor: theme.colors.separator,
      fontWeight: 600,
      backgroundColor: theme.colors.gray10
    },

    // Caret styles begin here.
    "& .header::after": {
      ...caret
    },

    "&:hover .header::after": {
      borderLeftColor: theme.colors.gray80
    },

    "&.open .header.open::after": {
      // rotate the caret to face down when an item is open.
      transform: "translate3d(-2px, 2px, 0) rotate(90deg)",
      borderLeftColor: theme.colors.gray80
    },

    // Spinner for async items replaces a caret.
    "&.updating .header::after": {
      width: 16,
      height: 16,
      border: 0,
      borderRadius: "50%",
      boxShadow: `1px 0px 0px 0px ${theme.colors.gray70} inset`,
      animation: `.7s ${spin} linear infinite`
    },

    "& .content": {
      position: "relative",
      paddingLeft: theme.spacing
    }
  }
}

class SidebarItem extends React.Component<IProps, IState> {
  state = {
    open: false,
    updating: false
  }

  componentWillMount() {
    this.setState({
      open: !!this.props.open
    })
  }

  async toggle() {
    if (!this.props.children) {
      return false
    }
    this.setState(() => ({ updating: true }))
    // If it is closed,
    if (typeof this.props.onClick === "function" && !this.state.open) {
      await this.props.onClick() // wait for the promise to resolve first.
    }
    this.setState(prevState => ({
      open: !prevState.open,
      updating: false
    }))
    return true
  }

  render() {
    /**
      Only the header should have a tooltip, else the tooltip will show
      even when the cursor is over the children... who may also have their
      own tooltips.
    */
    return (
      <div
        style={this.props.style}
        key={this.props.id}
        className={`${this.props.className} ${this.state.updating ? "updating" : ""} ${this.state.open ? "open" : ""}`}
      >
        <div className={`header ${this.state.open ? "open" : ""}`} onClick={() => this.toggle()}>
          {this.props.label}
        </div>
        {this.state.open ? <div className="content">{this.props.children}</div> : ""}
      </div>
    )
  }
}

export default glamorous(SidebarItem)(style)
export { SidebarItem }
