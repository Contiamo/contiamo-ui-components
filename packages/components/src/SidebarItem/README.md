A sidebar item can be thought of as a leaf node.
