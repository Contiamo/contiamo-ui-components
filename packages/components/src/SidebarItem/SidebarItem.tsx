import * as React from "react"
import styled from "react-emotion"
import { OperationalStyleConstants, Theme } from "@operational/theme"
import { readableTextColor, darken } from "@operational/utils"
import { WithTheme, Css, CssStatic } from "../types"
export interface Props {
  id?: string
  /** `css` prop as expected in a glamorous component */

  css?: Css
  className?: string
  children?: React.ReactNode
  onClick?: () => void
  /** Disabled */

  disabled?: boolean
  /** Active */

  active?: boolean
}
const Container = styled("div")(
  ({
    theme,
    isDisabled,
    isActive,
  }: {
    theme?: OperationalStyleConstants & {
      deprecated: Theme
    }
    isDisabled: boolean
    isActive: boolean
  }): CssStatic => ({
    label: "sidebaritem",
    backgroundColor: theme.deprecated.colors.white,
    height: 30,
    ...theme.deprecated.typography.body,
    fontWeight: isActive ? 600 : 400,
    position: "relative",
    borderBottom: "1px solid",
    borderColor: theme.deprecated.colors.separator,
    display: "flex",
    alignItems: "center",
    justifyContent: "space-between",
    padding: `0px ${theme.deprecated.spacing}px`,
    cursor: "pointer",
    textDecoration: "none",
    color: isActive ? theme.deprecated.colors.linkText : theme.deprecated.colors.text,
    ...(isDisabled
      ? {
          opacity: 0.25,
          pointerEvents: "none",
        }
      : {}),
    ":hover": {
      backgroundColor: theme.deprecated.colors.lighterBackground,
    },
    ":focus": {
      outline: 0,
    },
  }),
)

const SidebarItem = (props: Props) => (
  <Container
    id={props.id}
    css={props.css}
    onClick={props.onClick}
    className={props.className}
    isActive={!!props.active}
    isDisabled={!!props.disabled}
  >
    {props.children}
  </Container>
)

export default SidebarItem
