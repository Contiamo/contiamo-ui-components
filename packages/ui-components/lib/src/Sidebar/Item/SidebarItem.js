"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var glamorous_1 = require("glamorous");
var SidebarItem_style_1 = require("./SidebarItem.style");
var SidebarItem = /** @class */ (function (_super) {
    __extends(SidebarItem, _super);
    function SidebarItem(props) {
        var _this = _super.call(this, props) || this;
        _this.state = {
            open: _this.props && _this.props.open,
            updating: false
        };
        return _this;
    }
    SidebarItem.prototype.toggle = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.props.children) {
                            return [2 /*return*/, false];
                        }
                        this.setState(function () { return ({ updating: true }); });
                        if (!(typeof this.props.onClick === "function" && !this.state.open)) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.props.onClick()]; // wait for the promise to resolve first.
                    case 1:
                        _a.sent(); // wait for the promise to resolve first.
                        _a.label = 2;
                    case 2:
                        this.setState(function (prevState) { return ({
                            open: !prevState.open,
                            updating: false
                        }); });
                        return [2 /*return*/, true];
                }
            });
        });
    };
    SidebarItem.prototype.render = function () {
        var _this = this;
        /**
          Only the header should have a tooltip, else the tooltip will show
          even when the cursor is over the children... who may also have their
          own tooltips.
        */
        var HeaderWithTooltip = glamorous_1.Div;
        return (React.createElement("div", { className: this.props.className + " " + (this.state.updating ? "updating" : "") + " " + (this.state.open ? "open" : "") },
            React.createElement("div", { className: "header " + (this.state.open ? "open" : ""), onClick: function () { return _this.toggle(); } }, this.props.label),
            this.state.open ? React.createElement("div", { className: "content" }, this.props.children) : ""));
    };
    SidebarItem.defaultProps = {
        open: false
    };
    return SidebarItem;
}(React.Component));
exports.SidebarItem = SidebarItem;
exports.default = glamorous_1.default(SidebarItem)(SidebarItem_style_1.default);
//# sourceMappingURL=SidebarItem.js.map