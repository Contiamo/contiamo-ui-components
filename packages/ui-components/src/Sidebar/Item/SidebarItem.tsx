import * as React from "react"

import glamorous, { GlamorousComponent } from "glamorous"
import { Theme } from "contiamo-ui-theme"

import style from "./SidebarItem.style"

export interface IProps {
  id?: string | number
  style?: {}
  className?: string
  label?: string
  children?: any
  open?: boolean
  onClick?: () => Promise<any>
  tooltip?: string
}

export interface IState {
  open: boolean
  updating: boolean // async, woo!
}

class SidebarItem extends React.Component<IProps, IState> {
  state = {
    open: false,
    updating: false
  }

  componentWillMount() {
    this.setState({
      open: !!this.props.open
    })
  }

  async toggle() {
    if (!this.props.children) {
      return false
    }
    this.setState(() => ({ updating: true }))
    // If it is closed,
    if (typeof this.props.onClick === "function" && !this.state.open) {
      await this.props.onClick() // wait for the promise to resolve first.
    }
    this.setState(prevState => ({
      open: !prevState.open,
      updating: false
    }))
    return true
  }

  render() {
    /**
      Only the header should have a tooltip, else the tooltip will show
      even when the cursor is over the children... who may also have their
      own tooltips.
    */
    return (
      <div
        style={this.props.style}
        key={this.props.id}
        className={`${this.props.className} ${this.state.updating ? "updating" : ""} ${this.state.open ? "open" : ""}`}
      >
        <div className={`header ${this.state.open ? "open" : ""}`} onClick={() => this.toggle()}>
          {this.props.label}
        </div>
        {this.state.open ? <div className="content">{this.props.children}</div> : ""}
      </div>
    )
  }
}

export default glamorous(SidebarItem)(style)
export { SidebarItem }
