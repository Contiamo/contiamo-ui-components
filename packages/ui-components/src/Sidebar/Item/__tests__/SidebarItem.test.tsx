import * as React from "react"
import { shallow } from "enzyme"

import { SidebarItem } from "../SidebarItem"

describe("SidebarItem", () => {
  it("Should initialize", () => {
    const renderedComponent = shallow(<SidebarItem>Hola, compadre!</SidebarItem>)
    expect(renderedComponent).toMatchSnapshot()
    expect(renderedComponent.state().updating).toBe(false)
  })
  it("Should handle asynchronous requests before expanding", async () => {
    const myPromise = () => new Promise(resolve => setTimeout(() => resolve(true), 1)),
      renderedComponent = shallow(<SidebarItem onClick={() => myPromise()}>Hola, compadre!</SidebarItem>)

    await renderedComponent.instance().toggle()
    expect(renderedComponent.state().open).toBe(true)
  })
  it("Should toggle", () => {
    const renderedComponent = shallow(<SidebarItem>Hola, compadre!</SidebarItem>)

    expect(renderedComponent.state().open).toBe(false)
    renderedComponent.instance().toggle()
    expect(renderedComponent.state().open).toBe(true)
  })
  it("Shouldn't toggle if there are no children", () => {
    const renderedComponent = shallow(<SidebarItem />)

    expect(renderedComponent.state().open).toBe(false)
    renderedComponent.instance().toggle()
    expect(renderedComponent.state().open).toBe(false)
  })
})
