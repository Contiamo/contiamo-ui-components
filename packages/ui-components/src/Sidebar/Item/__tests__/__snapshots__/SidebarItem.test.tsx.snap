// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarItem Should initialize 1`] = `
<div
  className="undefined  "
>
  <div
    className="header "
    onClick={[Function]}
  />
</div>
`;
